### PR TITLE
Okx: Websocket order channel fixes

### PR DESCRIPTION
**Pull request by gbjk**
The driver for this PR was a complete failure to deserialise (issue 1).
In creating tests for those fixes, I came across other issues and fixed them at the same time.

- [x] Fixes float64 with string annotation erroring on empty strings, specifically px field in market orders:
     `Okx Order Push Data error json: invalid use of ,string struct tag, trying to unmarshal "" into float64`
- [x] Switch to convert.StringToFloat64 instead of okxNumericalValue
- [x]  Fix typo in Notional* field names; Ironically prevented them from erroring
- [x] Add tests
- [x] Populate ClientOrderID and FilledTime
- [x] Add UnmarshalJSON to order.Side and use throughout the WS (one exception for a nullable)
- [x] Add Fee and FeeAsset
- [x] Fix Amount, QuoteAmount, ExecutedAmount and RemainingAmount

There's a cherry-pick from #1333 in here, because I wanted to use TestFixtureToDataHandler,
and also a fractal of the ClientOrderId naming fix from the same PR.
I can rebase on top.
It also sets TestFixtureToDataHandler to be a t.Helper(), which feels right.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run
- [x] TestOrderPushData

### CODE REVIEW

File: exchanges/okx/okx_type_convert.go [hunk 1]
```diff
@@ -212,35 +210,21 @@ func (a *OrderDetail) UnmarshalJSON(data []byte) error {
 }
 
 // UnmarshalJSON deserializes JSON, and timestamp information.
-func (a *PendingOrderItem) UnmarshalJSON(data []byte) error {
+func (o *PendingOrderItem) UnmarshalJSON(data []byte) error {
```
Comment (shazbert) [inline exchanges/okx/okx_type_convert.go:213]: Can remove this method and it was obfuscating an issue with unmarshalling reduce only string. 
```diff
diff --git a/exchanges/okx/okx_type_convert.go b/exchanges/okx/okx_type_convert.go
index a937352a7..6b4e9afc3 100644
--- a/exchanges/okx/okx_type_convert.go
+++ b/exchanges/okx/okx_type_convert.go
@@ -209,24 +209,6 @@ func (a *OrderDetail) UnmarshalJSON(data []byte) error {
        return nil
 }

-// UnmarshalJSON deserializes JSON, and timestamp information.
-func (o *PendingOrderItem) UnmarshalJSON(data []byte) error {
-       type Alias PendingOrderItem
-       a := &struct {
-               *Alias
-               UpdateTime   okxUnixMilliTime `json:"uTime"`
-               CreationTime okxUnixMilliTime `json:"cTime"`
-               FillTime     okxUnixMilliTime `json:"fillTime"`
-       }{
-               Alias: (*Alias)(o),
-       }
-       err := json.Unmarshal(data, a)
-       o.CreationTime = a.CreationTime.Time()
-       o.UpdateTime = a.UpdateTime.Time()
-       o.FillTime = a.FillTime.Time()
-       return err
-}
-
 // UnmarshalJSON deserializes JSON, and timestamp information.
 func (a *RfqTradeResponse) UnmarshalJSON(data []byte) error {
        type Alias RfqTradeResponse
diff --git a/exchanges/okx/okx_types.go b/exchanges/okx/okx_types.go
index 559635069..be2adbe27 100644
--- a/exchanges/okx/okx_types.go
+++ b/exchanges/okx/okx_types.go
@@ -719,7 +719,7 @@ type OrderHistoryRequestParams struct {
 type PendingOrderItem struct {
        AccumulatedFillSize        convert.StringToFloat64 `json:"accFillSz"`
        AveragePrice               convert.StringToFloat64 `json:"avgPx"`
-       CreationTime               time.Time               `json:"cTime"`
+       CreationTime               okxUnixMilliTime        `json:"cTime"`
        Category                   string                  `json:"category"`
        Currency                   string                  `json:"ccy"`
        ClientOrderID              string                  `json:"clOrdId"`
@@ -727,7 +727,7 @@ type PendingOrderItem struct {
        FeeCurrency                currency.Code           `json:"feeCcy"`
        LastFilledPrice            convert.StringToFloat64 `json:"fillPx"`
        LastFilledSize             convert.StringToFloat64 `json:"fillSz"`
-       FillTime                   time.Time               `json:"fillTime"`
+       FillTime                   okxUnixMilliTime        `json:"fillTime"`
        InstrumentID               string                  `json:"instId"`
        InstrumentType             string                  `json:"instType"`
        Leverage                   convert.StringToFloat64 `json:"lever"`
@@ -752,7 +752,7 @@ type PendingOrderItem struct {
        TakeProfitTriggerPrice     convert.StringToFloat64 `json:"tpTriggerPx"`
        TakeProfitTriggerPriceType string                  `json:"tpTriggerPxType"`
        TradeID                    string                  `json:"tradeId"`
-       UpdateTime                 time.Time               `json:"uTime"`
+       UpdateTime                 okxUnixMilliTime        `json:"uTime"`
 }

 // TransactionDetailRequestParams retrieve recently-filled transaction details in the last 3 day.
@@ -2499,7 +2499,7 @@ type WsBalanceAndPosition struct {

 // WsOrder represents a websocket order.
 type WsOrder struct {
-       PendingOrderItem
+       *PendingOrderItem
        AmendResult     string                  `json:"amendResult"`
        Code            string                  `json:"code"`
        ExecType        string                  `json:"execType"`
@@ -2508,7 +2508,7 @@ type WsOrder struct {
        FillNotionalUsd convert.StringToFloat64 `json:"fillNotionalUsd"`
        Msg             string                  `json:"msg"`
        NotionalUSD     convert.StringToFloat64 `json:"notionalUsd"`
-       ReduceOnly      bool                    `json:"reduceOnly"`
+       ReduceOnly      bool                    `json:"reduceOnly,string"`
        RequestID       string                  `json:"reqId"`
 }

diff --git a/exchanges/okx/okx_websocket.go b/exchanges/okx/okx_websocket.go
index 102f5d278..77f0f2312 100644
--- a/exchanges/okx/okx_websocket.go
+++ b/exchanges/okx/okx_websocket.go
@@ -1060,7 +1060,7 @@ func (ok *Okx) wsProcessOrders(respRaw []byte) error {
                        AssetType:            a,
                        AverageExecutedPrice: response.Data[x].AveragePrice.Float64(),
                        ClientOrderID:        response.Data[x].ClientOrderID,
-                       Date:                 response.Data[x].CreationTime,
+                       Date:                 response.Data[x].CreationTime.Time(),
                        Exchange:             ok.Name,
                        ExecutedAmount:       execAmount,
                        RemainingAmount:      remAmount,
@@ -1074,7 +1074,7 @@ func (ok *Okx) wsProcessOrders(respRaw []byte) error {
                        Type:                 orderType,
                }
                if orderStatus == order.Filled {
-                       d.CloseTime = response.Data[x].FillTime
+                       d.CloseTime = response.Data[x].FillTime.Time()
                        if d.Amount == 0 {
                                d.Amount = d.ExecutedAmount
                        }
```

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for updating this, very cool. One issue I noticed from testing.

File: exchanges/okx/okx_type_convert.go (hunk 1 above)
Comment (shazbert) [reply exchanges/okx/okx_type_convert.go:213]: That pointer to `PendingOrderItem` can be removed. Whooops.

Comment (gbjk) [reply exchanges/okx/okx_type_convert.go:213]: Okay. That's done.
ReduceOnly is part of non-SPOT, so I wasn't (yet) testing it.
I'll get to that soon.
In the meantime I've taken your changes.

**Review by shazbert — APPROVED** [review]
Thank you sir! 🚀

**Review by gloriousCode — CHANGES_REQUESTED** [review]
With https://github.com/thrasher-corp/gocryptotrader/pull/1333 merged, looks like you'll need to get those changes in. I'll continue to review avoiding the sharedtest changes etc

File: exchanges/okx/okx_websocket.go [hunk 2]
```diff
@@ -1044,20 +1039,47 @@ func (ok *Okx) wsProcessOrders(respRaw []byte) error {
...
+
+		if response.Data[x].SizeType == "quote_ccy" {
+			quoteAmount = response.Data[x].Size.Float64()
+		} else {
+			amount = response.Data[x].Size.Float64()
+		}
+
+		execAmount := response.Data[x].AccumulatedFillSize.Float64()
+
+		// We can only work out the
```
Comment (gloriousCode) [inline exchanges/okx/okx_websocket.go:1052]: Is this supposed to run onto `amount` below or is this comment incomplete?

**Review by gloriousCode — COMMENTED** [review]
Just one comment on a comment

File: exchanges/okx/okx_websocket.go (hunk 2 above)
Comment (gbjk) [reply exchanges/okx/okx_websocket.go]: Fixed during rebase.
Thanks for spotting

File: exchanges/order/order_test.go [hunk 3]
```diff
@@ -2045,3 +2047,13 @@ func TestAdjustQuoteAmount(t *testing.T) {
 		t.Fatalf("received: '%v' but expected: '%v'", s.Amount, 5.22222222)
 	}
 }
+
+func TestSideUnmarshal(t *testing.T) {
+	t.Parallel()
+	var s Side
+	assert.Nil(t, s.UnmarshalJSON([]byte(`"SELL"`)), "Quoted valid side okay")
+	assert.Equal(t, Sell, s, "Correctly set order Side")
+	assert.ErrorIs(t, s.UnmarshalJSON([]byte(`"STEAL"`)), errUnrecognisedOrderSide, "Quoted invalid side errors")
```
Comment (gloriousCode) [inline exchanges/order/order_test.go:2056]: `errUnrecognisedOrderSide` doesn't exist

`ErrSideIsInvalid` ?

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for the not-so-easy rebase! Looks good, one thing

File: exchanges/order/order_test.go (hunk 3 above)
Comment (gbjk) [reply exchanges/order/order_test.go]: Whoops. Sitting in my local branch unpushed 🤦 
Fixed 04e48924

**Comment by gbjk** [comment]
@shazbert Sorry. Something caused me to get distracted and not push.
Probably bears 🐻 It's usually bears.

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for making those changes

**Review by thrasher- — APPROVED** [review]
Great stuff, like the new order.Side unmarshaller too @gbjk !

**Pull request merged**